### DISCUSSION TITLE
Support for generating non python code

### DISCUSSION THREAD

**Discussion by andersesbensen** [koxudaxi/datamodel-code-generator#1657] [Ideas]
I’m looking for a way to generate C++ code and other code and documentation from json schemas. I saw that this project allows for custom Jinja templates. So I thought that I might be able to use your generator. However it seems that the way things are now, the templates must generate python code at a formatter is handling the generated code. Also field variable type_ hint is a derived pydantic type.

Would it be feasible to change the generator into a more generic generator which could be used to generate arbitrary transformations of JSON schemas? 

I want to make C++ code, Rust code, documentation and even UI’s from a schema.

Any thoughts on this?

**Comment by koxudaxi** [comment]
@andersesbensen 
> Would it be feasible to change the generator into a more generic generator which could be used to generate arbitrary transformations of JSON schemas?

Yes, if it is considered valuable enough.

I have been thinking about something similar for years.

First of all, I created this tool to create a `Pydantic` model from `OpenAPI`. However, I did not include `Pydantic` in the package name to allow for diversity in input and output formats.
As a result, it now supports Python [TypeDict](https://docs.python.org/3/library/typing.html#typing.TypedDict), [dataclasses](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) and [msgspec](https://github.com/jcrist/msgspec).

> However it seems that the way things are now, the templates must generate python code at a formatter is handling the generated code

From the beginning, formatting has been optional at the method level, so it can easily be disabled by command options or output format.

Of course, the module structure is different from Python, so some refactoring is required to support other languages.

> Also field variable type_ hint is a derived pydantic type.

This tool parses the schema and generates an abstract object called `datamodel`.
This is the relationship between the class name and the fields, and I think it is probably possible to generate objects that hold a similar structure, such as [kotlin's data classes](https://kotlinlang.org/docs/data-classes.html) or [TypeScript's Zod](https://github.com/colinhacks/zod).
Also, from the design phase, I did not keep datamodel as a Python AST or CST in order to preserve the possibility of outputting non-Python code.

This project started as a small tool for me, and even after the number of users increased, I did not do any major refactoring for compatibility.
For example, I am thinking that it might be nice to rewrite it in Rust for execution speed and a single binary. However, it is not clear that the code generator needs that much performance, and since it provides a docker image, I may not feel that not having a single binary is a disadvantage in many use cases.
Besides, I believe it gets a lot of pull requests by being written in Python; I can't really imagine many people casually giving pull requests to Rust's complex implementation.
Even if you are not a Python user, it is far easier to understand Python syntax than Rust.

What do you think of the above thoughts?

**Reply by andersesbensen** [reply]
Nice to know that we share the same vision :-),

Regarding rust, I don't think there is much gained in porting it to rust, as I see this as a compile-time tool so performance as you say is not much of an issue. And I agree that the Python user base is much bigger than Rust at the moment.

Without quite understanding the details of your code, it looks like it would not be that difficult to separate the Python parts from the generic data-model parts. 

Make sure that the Jinja scripts has access to basic traits for each model property, like

is_optional
is_list
is_enum
default
required
abstract_type_hint -- this would be more or less like the type_hint, so instead of List[Foo] or List[Boolean] , it would just return Foo or Boolean.

I think you already have much of this working already.

**Reply by koxudaxi** [reply]
Thank you for your comment.

> Regarding rust, I don't think there is much gained in porting it to rust, as I see this as a compile-time tool so performance as you say is not much of an issue. And I agree that the Python user base is much bigger than Rust at the moment.

Some users want to use the library in runtime. 
However, I am not sure how beneficial that would be.

> Make sure that the Jinja scripts has access to basic traits for each model property, like

We can do it via `DataModelField`
The design policy is to implement properties, etc., in the model `DataModel` and `DataModelField` classes for access to them in order to avoid complex template descriptions and to allow for flexible processing.

https://github.com/koxudaxi/datamodel-code-generator/blob/2a3d27df822b6a8c3ce11fda8e15a71b42a7f579/datamodel_code_generator/model/typed_dict.py#L46-L112
https://github.com/koxudaxi/datamodel-code-generator/blob/2a3d27df822b6a8c3ce11fda8e15a71b42a7f579/datamodel_code_generator/model/typed_dict.py#L115-L149

The part expects only Python structure. we should abstract the part as a method for every output file type.
https://github.com/koxudaxi/datamodel-code-generator/blob/2a3d27df822b6a8c3ce11fda8e15a71b42a7f579/datamodel_code_generator/parser/base.py#L1064-L1201

Also, `ModelResolver` has the logic to generate valid Python names like class names, and field names...
https://github.com/koxudaxi/datamodel-code-generator/blob/2a3d27df822b6a8c3ce11fda8e15a71b42a7f579/datamodel_code_generator/reference.py#L211-L264

It seems to me that a simple output type could be implemented relatively quickly.
If we tackle this, which output file type would be appropriate?

**Reply by andersesbensen** [reply]
I think it would be nice to do a proof of concept with a plain C++ struct generator, which can just hold data. 


So an output could be
```cpp
enum Pet {
  Fish,
  Dog,
 Cat,
};


/// Description
/// bla bal
struct Person {
  /// Name of the person
  /// 
  std::string name;

  /// Set if the person has hair
  ///
  std::optional<std::string> hair_color;
  
  /// Age
  ///
  int age;

  /// Height in meters
  ///  
 double height;

  std::vector<Pet> pets;
}
```

I think here we should use the C++ containers std::optional and std::vector to hold optional values and arrays. I think the Python get_valid_name applies to more or less all languages I can think of.

Could the JSON to C++ primitive type map be a template in itself?

Something like 
typemap.jinja:
```jinja
{%- if is_number %}
double
{%- endif %}
{%- if is_integer %}
int
{%- endif %}
{%- if is_string %}
std::string
{%- endif %}
{%- if is_class %}
{{name}}
{%- endif %}
```

And a container map
containermap.jinja:

```jinja
{%- if is_array %}
std::vector<{{name}}>
{%- endif %}
{%- if is_optional %}
std::optional<{{name}}>
{%- endif %}
{%- if is_anyof %}
std::variant<
{% for type in type_list-%}
{{ type }},
{% endfor -%}
>
{%- endif %}

```

I don't know if this to too complicated, but this would allow all language-specific files in the template directory. 

Regarding the formatter, I would argue that it should not be part of the generation at all, as the formatter is language specific, for rust code, I would use rustfmt, for C++ code I would use clang-format and for markdown, a third tool would be used.

**Reply by koxudaxi** [reply]
@andersesbensen 

> I think the Python get_valid_name applies to more or less all languages I can think of.

OK, it's good.

> Could the JSON to C++ primitive type map be a template in itself?

Interesting suggestion.
I feel it would work well for primitive types and no default value.
However, I am not sure if it is possible to express everything in the template, since the output often refers to another model or involves multiple elements, such as `Optional[Union[List[Optional[str]], int, float]]`. (Even if it could, I fear that it would become so complex that no one would be able to understand it.)
However, I agree with the idea of having a property as an open interface.

> Regarding the formatter, I would argue that it should not be part of the generation at all, as the formatter is language specific, for rust code, I would use rustfmt, for C++ code I would use clang-format and for markdown, a third tool would be used.

I agree. the cli provides the formatter for Python by default, but it's ok it may not provide formatter  for other languages.
Of course, calling `rustfmt` or similar in a sub-process as an option is not a bad idea.

**Reply by andersesbensen** [reply]
I'm trying to play around with the code and there is something I don't quite understand, If I add
```python
    @property
    def foo(self) -> str:
        return self.data_type.is_list
```
to DataModelFieldBase and use print the field in a template like this:
```` 
    {%- for field in fields %}
        {{ field.foo }}
    {%- endfor -%}
````
It always returns False?

**Reply by koxudaxi** [reply]
@andersesbensen 
Can you share the example input?

**Reply by andersesbensen** [reply]
This is my model

```json
{
    "$id": "https://example.com/person.schema.json",
    "$schema": "http://json-schema.org/draft-04/schema#",
    "title": "ProjectSettings",
    "type": "object",
    "description": "Main project settings",

    "properties": {
        "batchmode": {
            "type": "boolean",
            "default": "false",
            "description": "Run though the calculations without stopping. This is typically for command line usage"
        },
        "import_settings": {
            "type"  :"array",
            "items": {
                "$ref": "#/definitions/ImportSettings"
            }
        }
    },
    "required": [
        "import_settings"
    ],
    "definitions": {
        "ImportSettings": {
            "type": "object",
            "properties": {
                "file_name": {
                    "type": "string"
                }
            }
        }
    }
}
```

**Reply by koxudaxi** [reply]
@andersesbensen 
I guess you can access the attribute by `self.data_type.data_types[0].is_list`
We should refactor the case.
https://github.com/koxudaxi/datamodel-code-generator/blob/2a3d27df822b6a8c3ce11fda8e15a71b42a7f579/datamodel_code_generator/types.py#L448-L449
